### PR TITLE
Add Artifactory build to the matrix

### PR DESCRIPTION
**Pull request by jaymalasinha**
Signed-off-by: Jaymala Sinha <jaymala@hashicorp.com>

### CODE REVIEW

File: enos/enos-scenario-autopilot.hcl [hunk 1]
```diff
@@ -78,11 +90,8 @@ scenario "autopilot" {
   }
 
   step "create_vault_cluster" {
-    module = module.vault_cluster
-    depends_on = [
-      step.create_vpc,
-      step.build_vault,
```
Comment (rebwill) [inline enos/enos-scenario-autopilot.hcl:84]: Out of curiosity, why are we no longer depending on `step.build_vault`?

Comment (jaymalasinha) [reply enos/enos-scenario-autopilot.hcl:84]: That's a great question.  In my earlier tests I thought I saw it fail unless I explicitly depended it, but with later versions of Enos I think the implicitly dependency when called within the step works, so I don't need to make it explicitly define depends_on anymore.  @ryancragun can confirm for us, if removing it is a good idea :) !

Comment (rebwill) [reply enos/enos-scenario-autopilot.hcl:84]: Last I heard we still needed to explicitly include the `depends_on`, unless there's something I missed...

File: .github/enos-run-matrices/ent.json [hunk 2]
```diff
@@ -1,23 +1,23 @@
...
       "aws_region": "us-west-2"
     },
     {
-      "scenario": "upgrade backend:raft consul_version:1.11.7 distro:rhel seal:shamir arch:amd64 builder:crt edition:ent",
+      "scenario": "upgrade backend:raft consul_version:1.11.7 distro:rhel seal:shamir arch:amd64 build:crt edition:ent",
       "aws_region": "us-west-1"
     },
     {
-      "scenario": "upgrade backend:consul consul_version:1.11.7 distro:rhel seal:awskms arch:amd64 builder:crt edition:ent",
+      "scenario": "upgrade backend:consul consul_version:1.11.7 distro:rhel seal:awskms arch:amd64 build:crt edition:ent",
```
Comment (ryancragun) [inline .github/enos-run-matrices/ent.json:16]: I dunno, I think I like the old way better. It communicates that we're choosing _how_ something is going to be built, rather than _which_ build we're choosing. But there's the problem: this variant describes both behaviors. If we're going to change it we probably need to think of a better word that communicates `This determines how the vault binary will be built or where it will be provided from`

Comment (ryancragun) [reply .github/enos-run-matrices/ent.json:16]: Maybe `artifact_source` or `source`?

Comment (ryancragun) [reply .github/enos-run-matrices/ent.json:16]: Curious about your thoughts in particular @rebwill since you're the most articulate of all of us.

File: enos/enos-scenario-autopilot.hcl [hunk 3]
```diff
@@ -38,13 +39,24 @@ scenario "autopilot" {
...
   step "build_vault" {
-    module = matrix.builder == "crt" ? module.build_crt : module.build_local
+    module = "build_${matrix.build}"
 
     variables {
-      build_tags  = var.vault_local_build_tags != null ? var.vault_local_build_tags : local.build_tags[matrix.edition]
-      bundle_path = local.bundle_path
-      goarch      = matrix.arch
-      goos        = "linux"
+      build_tags            = var.vault_local_build_tags != null ? var.vault_local_build_tags : local.build_tags[matrix.edition]
```
Comment (ryancragun) [inline enos/enos-scenario-autopilot.hcl:45]: I see why you want `try()`. Maybe we should cut an enos release before we merge this :D

File: enos/enos-scenario-autopilot.hcl [hunk 4]
```diff
@@ -38,13 +39,24 @@ scenario "autopilot" {
...
+      bundle_path           = local.bundle_path
+      goarch                = matrix.arch
+      goos                  = "linux"
+      artifactory_host      = matrix.build == "artifactory" ? var.artifactory_host : null
+      artifactory_repo      = matrix.build == "artifactory" ? var.artifactory_repo : null
```
Comment (ryancragun) [inline enos/enos-scenario-autopilot.hcl:50]: These all have `null` default values. Is it necessary to check for the matrix build type for all of them if they're going to be null or set by the caller?

File: enos/enos-scenario-autopilot.hcl [hunk 5]
```diff
@@ -78,11 +90,8 @@ scenario "autopilot" {
   }
 
   step "create_vault_cluster" {
-    module = module.vault_cluster
-    depends_on = [
-      step.create_vpc,
-      step.build_vault,
-    ]
+    module     = module.vault_cluster
+    depends_on = [step.create_vpc]
```
Comment (ryancragun) [inline enos/enos-scenario-autopilot.hcl:94]: This can be racy. If we're using a local build variant or artifactory it's possible we try to install vault on the machines before the artifact has been built or downloaded. We should still depend on `build_vault`. Pretty sure we actually saw that happen and thus the dependency was added.

File: enos/enos-scenario-autopilot.hcl [hunk 6]
```diff
@@ -141,6 +151,7 @@ scenario "autopilot" {
       vault_init                  = false
       vault_license               = step.read_license.license
       vault_local_artifact_path   = local.bundle_path
+      vault_artifactory_release   = local.install_artifactory_artifact ? step.build_vault.vault_artifactory_release : null
```
Comment (ryancragun) [inline enos/enos-scenario-autopilot.hcl:154]: I assume the other build modules will always return null for the artifactory release so we can probably always use that output.

**Review by ryancragun — COMMENTED** [review]
The bulk of the feedback here applies to all three scenarios.

File: enos/enos-scenario-autopilot.hcl (hunk 1 above)
Comment (ryancragun) [reply enos/enos-scenario-autopilot.hcl:84]: https://github.com/hashicorp/vault/pull/17353#discussion_r988275001

File: .github/enos-run-matrices/ent.json (hunk 2 above)
Comment (rebwill) [reply .github/enos-run-matrices/ent.json:16]: Hmm. Now that I think about it, the name of the `crt` variant is actually kind of a misnomer, because that variant gets the artifact from our `build` workflow, which lives in the repo and really has nothing to do with CRT other than triggering the CRT workflows when the build workflow runs on merge to main.

What about the following: we call the matrix item `artifact_source`, and the variants are the following: 

* `local_build` -- Build the artifact locally
* `CI_build` -- Receive the artifact from the Build workflow in CI
* `artifactory` -- Receive the artifact from Artifactory

Comment (ryancragun) [reply .github/enos-run-matrices/ent.json:16]: Yeah, I chose CRT to convey to people unfamiliar that the build would come _from_ crt. But in both the local and CI build cases we're expecting a local artifact, so I'm confused with the `local`. I'm hesitant to make things too wordy so I'll provide two more suggestions.

`artifact_source:{build_current_branch, local_artifact, download_artifactory}`
`source:{current_branch, local_artifact, artifactory}`

Comment (rebwill) [reply .github/enos-run-matrices/ent.json:16]: Fair. Of those two, I'd go with the first one.

Comment (ryancragun) [reply .github/enos-run-matrices/ent.json:16]: Is there a middle ground?

File: enos/enos-scenario-autopilot.hcl (hunk 3 above)
Comment (ryancragun) [reply enos/enos-scenario-autopilot.hcl:45]: (note: I just released 0.0.12 with `try()`)

File: .github/enos-run-matrices/ent.json (hunk 2 above)
Comment (joshbrand) [reply .github/enos-run-matrices/ent.json:16]: the CRT/build confusion has definitely caught me off guard in the past before, I think `artifact_source` is definitely a clearer name, but that then makes `local_artifact` redundant/confusing somewhat? What about `artifact_source:use_local` or similar?

Comment (ryancragun) [reply .github/enos-run-matrices/ent.json:16]: `artifact_source:{build,local,artifactory}` ?

Comment (jaymalasinha) [reply .github/enos-run-matrices/ent.json:16]: :( words are hard, in this case I think crt makes better sense to me because I know the artifact is coming from CRT build, while local is it is being built locally.  The branch in both cases will always be current branch, no 😕 ?

File: enos/enos-scenario-autopilot.hcl (hunk 4 above)
Comment (jaymalasinha) [reply enos/enos-scenario-autopilot.hcl:50]: 💭 👍

File: .github/enos-run-matrices/ent.json (hunk 2 above)
Comment (jaymalasinha) [reply .github/enos-run-matrices/ent.json:16]: `artifact_source:{ci,local,artifactory}` ?

Comment (jaymalasinha) [reply .github/enos-run-matrices/ent.json:16]: may be just `source` instead of `artifact_source` keeps it short ?

Comment (jaymalasinha) [reply .github/enos-run-matrices/ent.json:16]: `crt` also builds from the current branch.  I realized the vagueness of `source` when I started writing it, so I updated it to `artifact_source = ["local", "ci", "artifactory"]`.  Does that look like a middle ground :D ?

File: enos/enos-scenario-autopilot.hcl (hunk 6 above)
Comment (jaymalasinha) [reply enos/enos-scenario-autopilot.hcl:154]: The other modules do not return that output, so might error out if I don't condition it here.

File: enos/enos-scenario-autopilot.hcl (hunk 5 above)
Comment (jaymalasinha) [reply enos/enos-scenario-autopilot.hcl]: Added back

File: enos/enos-scenario-autopilot.hcl (hunk 3 above)
Comment (jaymalasinha) [reply enos/enos-scenario-autopilot.hcl]: updated to use try

File: enos/modules/vault_artifactory_artifact/variables.tf [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+  description = "The artifactory host to search for vault artifacts"
+  default     = "https://artifactory.hashicorp.engineering/artifactory"
+}
+
+variable "artifactory_repo" {
+  type        = string
+  description = "The artifactory repo to search for vault artifacts"
+  default     = "hashicorp-crt-stable-local*"
+}
+# variable "ami_id" {}
```
Comment (joshbrand) [inline enos/modules/vault_artifactory_artifact/variables.tf:26]: should we remove these if they're not needed (for clarity?)

File: enos/modules/vault_artifactory_artifact/main.tf [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+  repo     = var.artifactory_repo
+  path     = var.edition == "oss" ? "vault/*" : "vault-enterprise/*"
+  properties = var.edition == "oss" ? tomap({
+    "commit"          = var.revision
+    "product-name"    = "vault"
+    "product-version" = var.vault_product_version
+    }) : tomap({
+    "commit"          = var.revision
+    "product-name"    = "vault-enterprise"
+    "product-version" = "${var.vault_product_version}+ent"
```
Comment (joshbrand) [inline enos/modules/vault_artifactory_artifact/main.tf:24]: this formatting feels weird, is this what `terraform fmt` spit out?

**Review by joshbrand — CHANGES_REQUESTED** [review]
just a few nits!

**Review by emilymianeil — APPROVED** [review]
lgtm (from CRT perspective)

**Comment by rebwill** [comment]
One thought regarding the `enos-run-matrices`: we will also need to add matrices for the nightly workflows. I am thinking we may want to rename the existing ones to include the `PR` prefix (e.g. `PR-artifactory-ent.json`), so that we can also add `nightly-artifactory-ent.json` and `nightly-artifactory-oss.json`.

File: .release/ci.hcl [hunk 9]
```diff
@@ -175,6 +175,18 @@ event "verify" {
   }
 }
 
+event "enos-verify-stable" {
+    depends = ["verify"]
+      action "enos-verify-stable" {
```
Comment (ryancragun) [inline .release/ci.hcl:180]: The formatting is strange for this event

File: .github/workflows/enos-verify-stable.yml [hunk 10]
```diff
@@ -0,0 +1,17 @@
+name: enos-verify-stable
```
Comment (ryancragun) [inline .github/workflows/enos-verify-stable.yml:1]: Am I correct in assuming that this workflow is triggered by CRT when something is promoted between "unstable" and "stable", i.e. after a push to main and the build workflow is run? If that's correct, we're doing almost the same validation of the same artifacts twice, except that validation is happening before and after security scanning, notorization, etc. Perhaps we ought not execute the enos-run scenario on pushes to main as we're now hooking into the pipeline at a later (better) step in the pipeline.

File: enos/modules/vault_artifactory_artifact/main.tf (hunk 8 above)
Comment (ryancragun) [reply enos/modules/vault_artifactory_artifact/main.tf:24]: The enos-fmt job ought to validate this, no?

File: .github/workflows/enos-verify-stable.yml (hunk 10 above)
Comment (jaymalasinha) [reply .github/workflows/enos-verify-stable.yml:1]: This is triggered after the verify artifact stage and runs Enos scenarios on the artifact in `stable` .  The scenario matrix run with this, tests more variants than the one run on PR.  We can add or reduce the number of scenarios to test based on this iteration.  After the artifact in stable is validated against these scenarios it should be ready to promote.  Let me know if this looks good to you as the first iteration of it.

File: enos/modules/vault_artifactory_artifact/main.tf (hunk 8 above)
Comment (jaymalasinha) [reply enos/modules/vault_artifactory_artifact/main.tf]: It did, and did not complain

File: .release/ci.hcl (hunk 9 above)
Comment (jaymalasinha) [reply .release/ci.hcl:180]: Yeah good catch.  My editor had 4space setting then.

**Review by joshbrand — APPROVED** [review]
